Feature steps model email template

YangengChen opened the pull request.

YangengChen committed:
Step and email template models

YangengChen committed:
Minor fix to routes and validation decorator

YangengChen committed:
Added Steps Route/Modified get one campaign request

YangengChen committed:
Updated ValidatorDecorator

benjaminjenkins1 commented on server/app.py:
```diff
@@ -31,8 +31,7 @@
 api.add_resource(GmailResources.GetGmailAddress, '/gmail/get_address')
-api.add_resource(CampaignResources.NewCampaign, '/campaigns')
 api.add_resource(CampaignResources.CampaignProspects, '/campaign/<int:id>/prospects')
-api.add_resource(CampaignResources.AddProspectsToCampaign, '/prospects/campaign')
```
Should these resources still be there?

YangengChen replied:
No, some of these resources have been merged with others.

benjaminjenkins1 commented on server/resources/CampaignResources.py:
```diff
@@ -39,17 +27,20 @@ def get(self, id):
-class AddProspectsToCampaign(Resource):
+
     @jwt_required
-    @validate_args("prospect_ids", "campaign_id")
```
The validate_args decorator is cool, but but I think we might be better off using the built-in validation in reqparse. For the modules that have multiple resources, you can have multiple parsers, which would make it more clear what the required and optional parameters are for each resource. With all of the arguments in one parser and a validator that takes any number of those, it would be hard to tell if there are optional args without looking through the whole method to see what args are used. add_argument has the optional required parameter, as well as ways to add error messages.
[https://flask-restful.readthedocs.io/en/0.3.5/reqparse.html#error-handling](https://flask-restful.readthedocs.io/en/0.3.5/reqparse.html#error-handling)

YangengChen replied:
I tried looking into the built-in validation in reqpars, but it seems that if you set a global parser in one file and make one of the arguments required, all the post methods in that file are required to have that argument as an input. Unless we can make multiple parser somehow and have them separate, what's your method of doing so?

benjaminjenkins1 replied:
Yeah, you can just make more than one RequestParser. Like you could have new_campaign_parser and add_prospects_parser that each have their own set of arguments, and just call parse_args() on the one you need in that method.

YangengChen replied:
Ah right, I see. But idk if Shums likes that the fact we have to spam code, idk we can bring it up during today's standup.

YangengChen committed:
Created a RequestParserGenerator

created a requestparsergenerator that auto generates a reqparser given specified fields

YangengChen committed:
Merge branch 'dev' into feature-steps-model-email-template
